design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+> thought. This may require downtime for applications that rely on the feature.
+
+Under this proposal all MRs are namespaced, and XRs may be namespaced or cluster
+scoped. A namespaced XR may only be composed of other namespaced resources (e.g.
+XRs, MRs, or any Kubernetes resource). Cluster scoped XRs may be composed of
+either cluster scoped or namespaced resources.
+
+This means a Usage may need to track the following relationships:
+
+* Namespaced type uses namespaced type
+* Cluster scoped type uses cluster scoped type
+* Cluster scoped type uses namespaced type
+
+Therefore I believe the Usage type must have both namespaced and cluster scoped
+variants. It’s subject to the same issue as EnvironmentConfigs - the Usage type
+is currently cluster scoped. Renaming it to ClusterUsage would be a breaking
+change.
+
+Like EnvironmentConfigs, I believe the options for Usage are either:
+
+* Introduce the namespaced variant as NamespacedUsage
+* Make a breaking API change
+
+Usages are currently an alpha feature, but at the time of writing code is
+committed to promote them to beta in Crossplane v1.19.
+
+### Promoting APIs to v2
+
+This document proposes incrementing the CompositeResourceDefinition API to v2,
+with the v2 API containing small changes from v1 - i.e. different default values
+and different allowable values.
+
+There’s no technical reason to promote any other core API to v2 - no changes are
+proposed to these APIs. However if Crossplane as a whole becomes v2 it may make
+sense to add identical v2 API variants of all core APIs for consistency.
+
+There’s less of a clear path to promote MR APIs to v2. Maybe MR APIs are still
+in v1betaX. MR API versions have not historically been coupled to the version or
+stability of core Crossplane, but instead reflect the maturity of the underlying
+provider and/or individual MR implementation.
+
+## Future Improvements
+
+This proposal opens a path for the following potential improvements.
+
+### Per-namespace Providers
+
+A Crossplane provider is ultimately a controller manager - a binary running a
+series of MR controllers. The package manager deploys these controller managers
+as Kubernetes pods.
+
+Currently Crossplane deploys a single pod per provider to the
+`crossplane-system` namespace. This pod reconciles all cluster scoped MRs.
+
+Under this proposal Crossplane would continue to deploy a single provider pod to
+the `crossplane-system` namespace. This pod would reconcile all MRs in any
+namespace. This is a common pattern in Kubernetes. A controller pod must be
+deployed to a specific namespace but typically reconciles custom resources in
+any namespace.
+
+It’s common for providers to use their pod identity to authenticate to the
+external system they reconcile (e.g. a cloud provider like AWS). Currently
+running a single provider pod for an entire control plane means the entire
+control plane must use that single pod’s identity to authenticate.
```

Comment:
In Azure, you can set multiple identities to the same pod, allowing it to select the right one at runtime by providing the ID. This is explained [here](https://azure.github.io/azure-workload-identity/docs/faq.html#how-to-federate-multiple-identities-with-a-kubernetes-service-account).
I don't feel the need to have namespaced providers. Having providers be able to horizontally scale though would be a much more appreciated feature in my opinion, but I know it's a different beast altogether.

Reply:
Technically this would be one way of horizontally scaling, it's just sharding the load by namespace.  Maybe not very efficient but it would be effective in some scenarios.

Reply:
Whilst I can see the benefits of both approaches, I kind of think the point about namespace sharding not being very efficient is quite poignant. 

Some namespaces will naturally end up with many more MRs than others which would leave the providers inside that namespace heavily loaded whilst other providers on the platform are under utilised. Thus namespace sharding is only efficient when there is a balance in the number of MRs across all namespaces.

Whilst it would obviously inherit some technical complexity, one approach might be to offer the ability to choose which scaling/sharding strategy was to be used when crossplane is first deployed into the cluster and allow the end user to choose the strategy that works best inside their environment?

Reply:
Interesting discussion here and feedback is heard. That said, I don't think we're going to bring provider sharding / horizontal scaling into scope for v2. At least not v2.0. 😉 So I'm going to resolve this thread.